fe/src/main/java/org/apache/doris/plugin/PluginMgr.java

```diff
@@ -0,0 +1,222 @@
+    private final static Logger LOG = LogManager.getLogger(PluginMgr.class);
+
+    private final Map<String, PluginLoader>[] plugins;
+
+    private String pluginDir;
+
+    public PluginMgr() {
+        plugins =  new Map[PluginType.MAX_PLUGIN_SIZE];
+
+        for (int i = 0; i < PluginType.MAX_PLUGIN_SIZE; i++) {
+            plugins[i] = Maps.newConcurrentMap();
+        }
+
+        this.pluginDir = Config.plugin_dir;
+    }
+
+    /**
+     * Dynamic install plugin thought install statement
+     */
+    public PluginInfo installPlugin(InstallPluginStmt stmt) throws IOException, UserException {
+        PluginLoader pluginLoader = new DynamicPluginLoader(Config.plugin_dir, stmt.getPluginPath());
+
+        try {
+            PluginInfo info = pluginLoader.getPluginInfo();
+
+            if (plugins[info.getTypeId()].containsKey(info.getName())) {
+                throw new UserException("plugin " + info.getName() + " has install.");
+            }
+
+            // install plugin
+            pluginLoader.install();
+
+            if (plugins[info.getTypeId()].putIfAbsent(info.getName(), pluginLoader) != null) {
+                throw new UserException("plugin " + info.getName() + " has install.");
+            }
+
+            return info;
+        } catch (IOException | UserException e) {
+            pluginLoader.uninstall();
+            throw e;
+        }
+    }
+
+    /**
+     * Dynamic uninstall plugin
+     */
+    public PluginInfo uninstallPlugin(String name) throws IOException, UserException {
+        for (PluginType type : PluginType.values()) {
+            if (plugins[type.ordinal()].containsKey(name)) {
+                PluginLoader loader = plugins[type.ordinal()].get(name);
+
+                if (null != loader) {
+                    // uninstall plugin
+                    loader.uninstall();
+                    plugins[type.ordinal()].remove(name);
+                    return loader.getPluginInfo();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * For built-in Plugin register
+     */
+    public boolean registerPlugin(PluginInfo pluginInfo, Plugin plugin) {
+        if (Objects.isNull(pluginInfo) || Objects.isNull(plugin) || Objects.isNull(pluginInfo.getType())
+                || Strings
+                .isNullOrEmpty(pluginInfo.getName())) {
+            return false;
+        }
+
+        PluginLoader loader = new BuiltinPluginLoader(Config.plugin_dir, pluginInfo, plugin);
+        PluginLoader checkLoader =
+                plugins[pluginInfo.getTypeId()].putIfAbsent(pluginInfo.getName(), loader);
+
+        return checkLoader == null;
+    }
+
+    /**
+     *
+     * Just load dynamic plugin
+     */
+    public void loadDynamicPlugin(PluginInfo info) throws IOException, UserException {
+        DynamicPluginLoader pluginLoader = new DynamicPluginLoader(Config.plugin_dir, info);
+
+        try {
+            // install plugin
+            pluginLoader.reload();
```

Comment:
If we reload failed, an exception will be thrown, and the plugin info will not be added to the `plugins`. So in replay process, it means we lost this plugin.

So I think even the plugin is reloaded failed, we should always put it in the `plugins`.

And each time we get a plugin from plugin manager, we will check if it is loaded successfully.